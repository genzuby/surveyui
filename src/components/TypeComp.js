import React from "react";
// for Adapted web design
import { isMobile, isBrowser } from "react-device-detect";
import { INPUT, INPUROTH, INPUROTHWEB } from "../styles/BodyStyles";
import "../styles/assets.scss";

// added styles for web device
const styles = {
  width: "calc(80% - 20vw)"
};

const TypeComp = ({ type, item, inputValue, setInputValue, setInputNext }) => {
  const onHandleCheckItem = e => {
    // If there is next value of belong to this input
    if (e.target.attributes.getNamedItem("data-next")) {
      setInputNext(e.target.attributes.getNamedItem("data-next").value);
    } else {
      setInputNext("");
    }
    // Set Inputvalue
    if (type === "Radio") {
      setInputValue(e.target.id);
    } else if (type === "Checkbox") {
      if (e.target.checked) {
        // when chceckbos selected, add value
        setInputValue([...inputValue, e.target.id]);
        // setInputValue(inputValue.push(e.target.id));
      } else {
        // when chceckbos un selected, delete value
        const arr = [...inputValue];
        arr.splice(arr.indexOf(e.target.id), 1);
        setInputValue(arr);
      }
    }
  };
  const renderCheckRadio = (val, inputType, groupName) => {
    return (
      <>
        <input
          type={inputType}
          name={groupName}
          id={val.value}
          checked={inputValue ? inputValue.includes(val.value) : false}
          data-next={val.next}
          onChange={onHandleCheckItem}
        />
        <label htmlFor={val.value}>{val.label}</label>
      </>
    );
  };

  const renderComp = () => {
    if (type !== "Page" && item && type) {
      const inputType = type.toLowerCase();
      const groupName = item.questionID;

      if (item.choices) {
        return item.choices.map((val, i) => {
          if (isMobile) {
            return (
              <INPUROTH key={i} data-test="typeCompEtcInput" type={inputType}>
                {renderCheckRadio(val, inputType, groupName)}
              </INPUROTH>
            );
          } else {
            return (
              <INPUROTHWEB
                key={i}
                data-test="typeCompEtcInput"
                type={inputType}
              >
                {renderCheckRadio(val, inputType, groupName)}
              </INPUROTHWEB>
            );
          }
        });
      } else if (inputType === "text") {
        if (isMobile) {
          return (
            <INPUT
              type={inputType}
              onChange={e => setInputValue(e.target.value)}
              value={inputValue}
              placeholder="Please enter your e-mail"
              data-test="typeCompTextInput"
            />
          );
        } else if (isBrowser) {
          return (
            <INPUT
              type={inputType}
              onChange={e => setInputValue(e.target.value)}
              style={styles}
              value={inputValue}
              placeholder="Please enter your e-mail"
              data-test="typeCompTextInput"
            />
          );
        }
      }
    }
  };

  return <div className="TypeComp">{renderComp()}</div>;
};

export default TypeComp;
