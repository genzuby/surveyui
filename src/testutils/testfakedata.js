export const surveylist = [
  {
    questionID: "page-welcome",
    type: "Page",
    heading: "Welcome to the survey",
    body: "You can start the survey by clicking the next button",
    next: "question-email"
  },
  {
    questionID: "page-uneligible",
    type: "Page",
    heading: "Not eligible",
    body: "We're sorry, but you are not eligible to use our producs/services"
  },
  {
    questionID: "page-finish",
    type: "Page",
    heading: "Thank you",
    body: "You have completed the survey"
  },
  {
    questionID: "question-email",
    type: "Text",
    question: "What is your email address?",
    next: "question-age"
  },
  {
    questionID: "question-age",
    type: "Radio",
    question: "Are you 18 or older?",
    choices: [
      {
        label: "Yes",
        value: "yes"
      },
      {
        label: "No",
        value: "no",
        next: "page-uneligible"
      }
    ],
    next: "question-problems"
  },
  {
    questionID: "question-problems",
    type: "Checkbox",
    question: "Do you suffer from any of the following problems?",
    choices: [
      {
        label: "I have dental implants",
        value: "dental_implants"
      },
      {
        label: "I have missing teeth",
        value: "missing_teeth"
      },
      {
        label: "I have unfilled cavities",
        value: "unfilled_cavities"
      },
      {
        label: "I have swollen or inflamed gums",
        value: "swollen_or_inflamed_gums"
      },
      {
        label: "I have retainers",
        value: "retainers"
      },
      {
        label: "None of the above",
        value: "no"
      }
    ],
    next: "page-finish"
  }
];

export const initData = [
  {
    surveyID: "345f52c7-78f6-4cc4-9f8a-0490744f7da3",
    surveyItems: [
      {
        questionID: "page-welcome",
        type: "Page",
        heading: "Welcome to the survey",
        body: "You can start the survey by clicking the next button",
        next: "question-email"
      },
      {
        questionID: "page-uneligible",
        type: "Page",
        heading: "Not eligible",
        body:
          "We're sorry, but you are not eligible to use our producs/services"
      },
      {
        questionID: "page-finish",
        type: "Page",
        heading: "Thank you",
        body: "You have completed the survey"
      },
      {
        questionID: "question-email",
        type: "Text",
        question: "What is your email address?",
        next: "question-age"
      },
      {
        questionID: "question-age",
        type: "Radio",
        question: "Are you 18 or older?",
        choices: [
          {
            label: "Yes",
            value: "yes"
          },
          {
            label: "No",
            value: "no",
            next: "page-uneligible"
          }
        ],
        next: "question-problems"
      },
      {
        questionID: "question-problems",
        type: "Checkbox",
        question: "Do you suffer from any of the following problems?",
        choices: [
          {
            label: "I have dental implants",
            value: "dental_implants"
          },
          {
            label: "I have missing teeth",
            value: "missing_teeth"
          },
          {
            label: "I have unfilled cavities",
            value: "unfilled_cavities"
          },
          {
            label: "I have swollen or inflamed gums",
            value: "swollen_or_inflamed_gums"
          },
          {
            label: "I have retainers",
            value: "retainers"
          },
          {
            label: "None of the above",
            value: "no"
          }
        ],
        next: "page-finish"
      }
    ]
  }
];

export const testAnswer = {
  userId: "test3333@gmail.com",
  userInfo: {
    under18: false,
    finished: true
  },
  answeredlist: {
    surveyID: "345f52c7-78f6-4cc4-9f8a-0490744f7da3",
    surveyItems: [
      {
        questionID: "question-problems",
        type: "Checkbox",
        choices: ["dental_implants", "unfilled_cavities"]
      }
    ]
  }
};
