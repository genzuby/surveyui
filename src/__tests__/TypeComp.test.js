import React from "react";
import { mount } from "enzyme";
import "../testutils/setup";
import { findByTestAttr } from "../testutils";

import TypeComp from "../components/TypeComp";

describe("Component of each types for display input JSX", () => {
  let wrapper;
  let type, item;

  beforeEach(() => {
    const props = {
      type,
      item
    };
    wrapper = mount(<TypeComp {...props} />);
  });

  describe("Should render without error", () => {
    it("renders correctly", () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render one input group", () => {
      const inputGroup = findByTestAttr(wrapper, "typeCompForm");
      expect(inputGroup.length).toBe(1);
    });
  });

  describe("Text Input", () => {
    const item = {
      questionID: "question-email",
      type: "Text",
      question: "What is your email address?",
      next: "question-age"
    };
    it("Should render input when type is Text ", () => {
      wrapper.setProps({ type: "Text", item });
      expect(wrapper.prop("type")).toEqual("Text");
      expect(wrapper.find("input")).toHaveLength(1);
    });
  });

  describe("Radio Input", () => {
    const item = {
      questionID: "question-age",
      type: "Radio",
      question: "Are you 18 or older?",
      choices: [
        {
          label: "Yes",
          value: "yes"
        },
        {
          label: "No",
          value: "no",
          next: "page-uneligible"
        }
      ],
      next: "question-problems"
    };

    it("Should render input when type is Radio ", () => {
      wrapper.setProps({ type: "Radio", item });
      expect(wrapper.prop("type")).toEqual("Radio");
      expect(wrapper.find("input")).toHaveLength(2);
    });
  });

  describe("Checkbox Input", () => {
    const item = {
      questionID: "question-problems",
      type: "Checkbox",
      question: "Do you suffer from any of the following problems?",
      choices: [
        {
          label: "I have dental implants",
          value: "dental_implants"
        },
        {
          label: "I have missing teeth",
          value: "missing_teeth"
        },
        {
          label: "I have unfilled cavities",
          value: "unfilled_cavities"
        },
        {
          label: "I have swollen or inflamed gums",
          value: "swollen_or_inflamed_gums"
        },
        {
          label: "I have retainers",
          value: "retainers"
        },
        {
          label: "None of the above",
          value: "no"
        }
      ],
      next: "page-finish"
    };

    it("Should render input when type is Radio ", () => {
      wrapper.setProps({ type: "Checkbox", item });
      expect(wrapper.prop("type")).toEqual("Checkbox");
      expect(wrapper.find("input")).toHaveLength(6);
    });
  });
});
