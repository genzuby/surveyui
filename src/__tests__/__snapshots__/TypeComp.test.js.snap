// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component of each types for display input JSX Should render without error renders correctly 1`] = `ReactWrapper {}`;
